scripts/solhint-custom/index.js

```diff
@@ -47,38 +50,63 @@ module.exports = [
     static ruleId = 'leading-underscore';
 
     VariableDeclaration(node) {
+      // TODO: do we want that rule ? Should no immutable variable have a prefix regardless of visibility ?
+      //
+      // else if (node.isImmutable) {
+      //   this.require(!node.name.startsWith('_'), node, 'Immutable variables should not have leading underscore');
+      // }
       if (node.isDeclaredConst) {
-        // TODO: expand visibility and fix
-        if (node.visibility === 'private' && /^_/.test(node.name)) {
-          this.error(node, 'Constant variables should not have leading underscore');
+        this.require(!node.name.startsWith('_'), node, 'Constant variables should not have leading underscore');
+      } else if (node.isStateVar) {
+        switch (node.visibility) {
+          case 'private':
+            this.require(node.name.startsWith('_'), node, 'Private state variables must have leading underscore');
+            break;
+          case 'internal':
+            this.require(node.name.startsWith('_'), node, 'Internal state variables must have leading underscore');
+            break;
+          case 'public':
+            this.require(!node.name.startsWith('_'), node, 'Public state variables should not have leading underscore');
+            break;
         }
-      } else if (node.visibility === 'private' && !/^_/.test(node.name)) {
-        this.error(node, 'Non-constant private variables must have leading underscore');
       }
     }
 
     FunctionDefinition(node) {
-      if (node.visibility === 'private' || (node.visibility === 'internal' && node.parent.kind !== 'library')) {
-        if (!/^_/.test(node.name)) {
-          this.error(node, 'Private and internal functions must have leading underscore');
-        }
-      }
-      if (node.visibility === 'internal' && node.parent.kind === 'library') {
-        if (/^_/.test(node.name)) {
-          this.error(node, 'Library internal functions should not have leading underscore');
-        }
+      switch (node.visibility) {
+        case 'external':
+          this.require(!node.name.startsWith('_'), node, 'External functions should not have leading underscore');
+          break;
+        case 'public':
+          this.require(!node.name.startsWith('_'), node, 'Public functions should not have leading underscore');
+          break;
+        case 'internal':
+          this.require(
+            node.name.startsWith('_') !== (node.parent.kind === 'library'),
+            node,
+            node.parent.kind === 'library'
+              ? 'Library internal functions should not have leading underscore'
```

Comment:
do we have a rule for non-internal library functions?

Reply:
non internal library function will be handeled like similar function in contracts:
- if public, line 81 applies and they must not have a `_` prefix
- if private, line 93 applies and they must have a `_` prefix

Reply:
private library function are definitelly ok.

Public ones ? We don't use them right now. I'm not sure if we want to forgid them. If we want to, I'd create a dedicated rule for that instead of pushing that into a "name format rule"

Reply:
added https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5794/commits/01d53b16c8c8f50eea92eb1ed0e4499c4280291a